ion-schema/src/isl/isl_type_reference.rs

```diff
@@ -208,20 +207,21 @@ impl IslTypeRefImpl {
                 if value_struct.get("id").is_none() {
                     let type_def= IslTypeImpl::from_owned_element(isl_version, value, inline_imported_types)?;
                     // if type reference contains `occurs` field and has modifier `$null_or` then return an error
-                    if nullability == NullOr && type_def.constraints()
-                        .iter()
-                        .any(|c| matches!(c, IslConstraintImpl::Occurs(_))) {
+                    if nullability == NullOr && value_struct.get("occurs").is_some() {
                         return invalid_schema_error(
                             "`$null_or` annotation is not supported for a type reference with an explicit `occurs` field",
                         )
                     }
 
-                    if !allow_occurs_field && isl_version == IslVersion::V2_0 && type_def.constraints()
-                        .iter()
-                        .any(|c| matches!(c, IslConstraintImpl::Occurs(_))) {
-                        return invalid_schema_error(
+                    if let Some(occurs_constraint) = value_struct.get("occurs") {
+                        if !allow_occurs_field && isl_version == IslVersion::V2_0 {
+                            return invalid_schema_error(
                             "A type reference with an explicit `occurs` field can only be used for `fields` and `ordered_elements` constraint",
-                        )
+                            )
+                        } else {
+                            // for ISL 1.0 ``occurs` field is a no op when used with constraints other than `fields` and `ordered_elements`
+                            IslTypeRefImpl::occurs_from_ion_element(occurs_constraint, isl_version)?;
```

Comment:
Is this here just to check for errors of some sort?

Reply:
Correct. Just to verify occurs is not zero/null or empty range.

Reply:
Is there any way it can be done in `IslVariablyOccurringTypeRef` instead?

Reply:
`IslVariablyOccurringTypeRef` will be created for constraints `fields` and `ordered_elements`. For any other constraints `IslTypeRef` will be created. For ISL 2.0 this will return an error but for ISL 1.0 we don't have the requirement to return an error but the `occurs` field related rules still counts, which is why I thought this will be the best place for it.
Why do you think this should be part of `IslVariablyOccurringTypeRef`?

Reply:
> Why do you think this should be part of `IslVariablyOccurringTypeRef`?

`IslVariablyOccurringTypeRef` is the type that actually cares/uses the `occurs` value, hence it seems appropriate there.

> For any other constraints IslTypeRef will be created. For ISL 2.0 this will return an error but for ISL 1.0 we don't have the requirement to return an error but the occurs field related rules still counts

So, in ISL 1.0, you're saying that occurs _is_ allowed in locations where it's basically just open content, but it still has to be a proper range in those locations? That's really weird, but if that's the case I guess it makes sense to check it here. I wonder if it would be better to just ignore `occurs` in wrong locations.

Reply:
> So, in ISL 1.0, you're saying that occurs is allowed in locations where it's basically just open content, but it still has to be a proper range in those locations? 
`occurs` can not be treated as open content as its part of the schema specification. But it can be used  in wrong locations with ISL 1.0 since it doesn't restrict `occurs` to be used with `fields` and `ordered_elements` only. 
This is definitely weird but its has to do with how ISL 1.0 accepts `occurs` in wrong locations but still doesn't consider it open content but rather serializes it as `occurs` and afterwards ignores it because its used in wrong location. 
Please let me know if I am misunderstanding this so that I can refactor accordingly.

Reply:
> Please let me know if I am misunderstanding this so that I can refactor accordingly.

No, you're not misunderstanding. I was just wistfully speculating that it would be nicer to implement if we could treat it as open content in places where it does nothing.

A little more info in the comment explaining why it is necessary to check `occurs` here might be helpful because other people might have the same question that I did.

Reply:
Got it. sure. I will add that.